crates/fj/src/abi/context.rs

```diff
@@ -23,7 +23,9 @@ impl<'a> From<&'a &dyn crate::models::Context> for Context<'a> {
             })) {
                 Ok(Some(arg)) => StringSlice::from_str(arg),
                 Ok(None) => StringSlice::from_str(""),
-                Err(payload) => crate::abi::on_panic(payload),
+                Err(payload) => StringSlice::from_str(
+                    crate::abi::on_panic(payload).to_string().as_str(),
```

Comment:
This is unsound: `on_panic` returns an `ffi_safe::String`, which is converted into a regular `String`, which is then dropped. The `StringSlice` that references it is returned from the function, meaning the the `StringSlice` outlives the `String` it references. Also see [`StringSlice`'s documentation](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj/src/abi/ffi_safe.rs#L259-L262).

---

In addition, this returns the error message just like it would return the value of the parameter. If I read the code right, this error message is then parsed as a value in the expanded model code ([[1]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L137), [[2]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L145)).

I think this is semi-okay in most cases, as it will cause a parse error. It might still be confusing to the user though. But it gets worse, if the parameter is a `String`. Imagine you're making a part that you want to engrave a custom message into, and you are engraving the error message instead. We're entering JavaScript territory here :smile: 

---

I think we can solve both issues by doing the following:

1. Return a `Result` (probably [`ffi_safe::Result`](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj/src/abi/ffi_safe.rs#L215-L221) is required) from this method. The `Err` variant can just be the `ffi_safe::String` returned from `on_panic`, for a start.
2. Adapt all methods that call this, as necessary, returning `Result` where needed. Specifically [`Context::get_argument`](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj/src/abi/context.rs#L39-L56) probably needs a custom error type, to distinguish between "argument not found" and "panicked" cases.
3. The generated model code already handles parse errors (here are the links to that expand code again: [[1]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L137), [[2]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L145)). We can adapt that code to handle errors resulting from a panic. ( I found it helpful to expand the macro code, for example with `cargo expand -p spacer` for the spacer model, to understand what's happening.)